app/services/file_attachment_blob_service.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+class FileAttachmentBlobService
+  attr_reader :file, :revision, :replacement_file
+
+  def initialize(file:, revision:, replacement_file: false)
+    @file = file
+    @revision = revision
+    @replacement_file = replacement_file
+  end
+
+  def blob_id
+    blob.id
+  end
+
+  def blob_filename
+    blob[:filename]
+  end
+
+  def number_of_pages
+    return unless file.content_type == "application/pdf"
+
+    PDF::Reader.new(file.tempfile).page_count
+  rescue PDF::Reader::MalformedPDFError, PDF::Reader::UnsupportedFeatureError, OpenSSL::Cipher::CipherError
+    nil
+  end
+
+private
+
+  def blob
+    @blob ||= ActiveStorage::Blob.create_after_upload!(
+      io: file,
+      filename: filename,
+      content_type: mime_type,
+    )
+  end
+
+  def mime_type
+    Marcel::MimeType.for(file,
+                         declared_type: file.content_type,
+                         name: file.original_filename)
+  end
+
+  def filename
+    return sanitized_filename if replacement_file
+
+    existing_filenames = revision.file_attachment_revisions.map(&:filename)
+    UniqueFilenameService.new(existing_filenames).call(file.original_filename)
```

Comment:
We could get rid of the need for the santized filename if both cases go through the unique filename service.

What I'd suggest is swapping the replacement_file from a boolean to a FileAttachment::Revision (and a name change)

then this method could be like:

```
def filename
  existing_filenames = revision.file_attachment_revisions.map(&:filename)
  existing_filenames.delete(replacement.filename) if replacement      
  UniqueFilenameService.new(existing_filenames).call(file.original_filename)
end
```

Reply:
Nice, added 👍